src/main/kotlin/uk/gov/justice/digital/hmpps/nomisprisonerapi/activities/PayRatesService.kt

```diff
@@ -72,20 +72,36 @@ class PayRatesService(
    * - any existing rates that are expired are retained
    * - any existing rates that are not included in the new request are expired
    */
-  fun buildNewPayRates(requestedPayRates: List<PayRateRequest>, existingActivity: CourseActivity): MutableList<CourseActivityPayRate> {
+  fun buildNewPayRates(requestedPayRates: List<PayRateRequest>, existingActivity: CourseActivity, previousActivityEndDate: LocalDate?): MutableList<CourseActivityPayRate> {
     val newPayRates = mutableListOf<CourseActivityPayRate>()
     val existingPayRates = existingActivity.payRates
     val tomorrow = LocalDate.now().plusDays(1)
 
     requestedPayRates.forEach { requestedPayRate ->
       val existingPayRate = existingPayRates.findExistingPayRate(requestedPayRate)
+
+      handleEndDate(previousActivityEndDate, existingActivity, existingPayRate)
+
       when {
-        existingPayRate == null -> newPayRates.add(requestedPayRate.toCourseActivityPayRate(existingActivity, getStartDate(requestedPayRate.incentiveLevel, requestedPayRate.payBand, existingPayRates, existingActivity.scheduleStartDate)))
+        existingPayRate == null -> newPayRates.add(
+          requestedPayRate.toCourseActivityPayRate(
+            existingActivity,
+            getStartDate(requestedPayRate.incentiveLevel, requestedPayRate.payBand, existingPayRates, existingActivity.scheduleStartDate),
+            existingActivity.scheduleEndDate,
+          ),
+        )
+
         existingPayRate.rateIsUnchanged(requestedPayRate) -> newPayRates.add(existingPayRate)
-        existingPayRate.rateIsChangedButNotYetActive(requestedPayRate) -> newPayRates.add(existingPayRate.apply { halfDayRate = requestedPayRate.rate }) // e.g. rate adjusted twice in same day
+
+        existingPayRate.rateIsChangedButNotYetActive(requestedPayRate) -> newPayRates.add(
+          existingPayRate.apply { halfDayRate = requestedPayRate.rate },
+        ) // e.g. rate adjusted twice in same day
```

Comment:
not clear what the comment is referring to

Reply:
Existing comment, but moved to more logical position